### ISSUE TITLE
theory improvement

### ISSUE THREAD

**Issue by GeneralShan** [somanchiu/ReSwapper#9]
so i have studyed your code.
it is very interesting as replication of INSWAPPER, but right now it is only distillation of inswapper. this will limit this model.
we can use this replication as initialization point for generator

we must need to use GAN architecture, because inswapper is derive from SimSwap.
We can use new SimSwapPlus architecture: https://www.scribd.com/document/784359864/SimSwap-Towards-Faster-and-High-Quality-Identity-Swapping

Inswapper original is base on simswap GAN. we can build replication with GAN of SimSwapPlus

**Comment by somanchiu** [comment]
@GeneralShan Thanks for reviewing my implementation and making some important PRs. I will review SimSwap as I haven't studied it deeply before.

As for the training code at the moment, you can say it is a "distillation" of InSwapper. If you look at it in the way deep learning works, it's about fitting the data. The current training code fits the outputs of InSwapper. It can also fit some deep fake images made by hand. As you can see, the training takes over 400,000 steps to converge. This means that we need a lot of "expected_output" images to train the model for higher resolution if we use the current training code. However, collecting that many hand-made deep fake images is very hard.

I also found that if I train the model with different resolutions images (128, 112, 96) at the same time, the generalization ability will be enhanced. We can observe this enhanced generalization ability from output resolutions of 64 x 64 to 1xx x 1xx. I'm wondering if it's possible to further enhance the generalization ability to higher resolutions like 256 x 256 or 512 x 512.

I agree with using GAN to train the model beacuse ImageA + ImageA source = ImageA, so "expected_output" is not necessary.

As for the architecture of the STYLE BLOCK and the overall model architecture, I created it through trial and error. If you take a close look at the InSwapper graph, you will see the STYLE BLOCK appears multiple times. It is obviously made with a loop. Exporting the model with opset_version=10 makes it easier to compare the graph in Netron. However, it will cause issue #8, which I will remark in README.md later.
Step 1. study the graph of InSwapper in NETRON and read the details from onnx.helper.printable_graph(model.graph)
Step 2. modify the model architecture in StyleTransferModel_128.py
Step 3. save the model in ONNX format so that I can compare it with the original InSwapper
Step 4. if the model is not the same as the original architecture, go back to step 1.

**Comment by GeneralShan** [comment]
good, i am performing similar experiment
i will training simswap Discriminator against ReSwapper architecture
i have read paper of both SIMSWAP and SIMSWAPPLUS

let us share results in this thread

**Comment by GeneralShan** [comment]
@somanchiu Thank you for detailed explanations of your process!!

I agree that GAN is the necessary directions for training

One difficulty I have see already:
- SimSWAPPLUS use VGGFace2-HQ, which is align FFHQ
- ArcFace include SimSwap ArcFace Checkpoint is align to InsightFACE

ID Retrieval for different alignment is lower, I observe

I think it will be good training on VGGFace2-HQ which is HIGH QUALITY dataset.
Also there is [EFHQ Dataset](https://vinairesearch.github.io/EFHQ/) which focus on pose variability
This includes face pose side profile and other pose that is not covered in FFHQ

I think best dataset will be combining VGGFace2-HQ and EFHQ augment with similar method

One more thing I see in NETRON:
- SimSwap use much smaller conv feature size of only 512, in encoder + Decoder stack, also in the STYLE BLOCK
- InSwapper use larger conv feature size of 1024

We all know InSwapper work much better compare to SimSwap. Most likely because this is the 2x model parameters.. and larger feature size in bottleneck in GENERATOR

**Comment by GeneralShan** [comment]
Also, if you want to have discussion, we can talk about the SIMSWAP paper and architecture
Please also look in SIMSWAPPLUS, and see that there is new lower-GFLOP operator, but also some change in layer count and model architecture

We shoud select best architecture.

I think it will be good direction to train SMALL model with SimSwap size, on 128px (with 0.5x parameter count compare to InSwapper)

If the training is stable and promising, we can train InSwapper size model and larger.

**Comment by somanchiu** [comment]
> @somanchiu
> 
> > This means that we need a lot of "expected_output" images to train the model for higher resolution if we use the current training code. However, collecting that many hand-made deep fake images is very hard.
> 
> If you meant by dataset from handmade df or sd. Then if this can help https://github.com/SelfishGene/SFHQ-dataset.
> 
> In Privacy section of readme, Since all images in this dataset are synthetically generated there are no privacy issues or license issues surrounding these images.

In the current training code, Not only "expected_output" is needed, we also need the target and the source image. "expected_output" created by target + source.

**Comment by somanchiu** [comment]
@GeneralShan I have created the [GAN branch](https://github.com/somanchiu/ReSwapper/tree/GAN). Do you have any updates on your experiment?

**Comment by GeneralShan** [comment]
i work local. i have limited time this week ,, i will share progress
i am making train run with SIMSWAP arcface FFHQ checkpoint as ID MODEL, and combine with SIMSWAP multiscale discriminator
- I will train subset of VGGFace2-HQ initial run

I will train smaller model variant of RESWAPPER architecture wth your STYLE BLOCK design. i will share the result when it is ready

I will also follow your work closely and I will contribute

**Comment by GeneralShan** [comment]
please share updat with me if you have train run stable

plealse also suggest use VGGFace2-HQ: https://github.com/NNNNAI/VGGFace2-HQ

Please ensure FFHQ alignment in use, because VGGFace2-HQ is FFHQ

Remember SIMSWAP ARCFACE is not align to FFHQ, which reduces performance, we may require finetune ARCFACE checkpoint to FFHQ align ID retrieval

**Comment by aesanchezgh** [comment]
VGGFace2-HQ is not a good dataset. It's artificial high quality. Nothing will substitute for unmodified high quality dataset. I am currently building one much better than FFHQ by hand. Do you guys have a Discord or Telegram where we can collaborate?

**Comment by somanchiu** [comment]
In my opinion, findings and knowledge should be shared in a way that everyone can easily access without needing an account. I don't see any problems with connecting everyone via GitHub issues or discussions. If you all think we should communicate outside of GitHub, then we'll move. What do you all think?

**Comment by aesanchezgh** [comment]
I agree in principle. However real time chat is easier for collaboration on details that don't have to flood this forum.

**Comment by GeneralShan** [comment]
> VGGFace2-HQ is not a good dataset. It's artificial high quality. Nothing will substitute for unmodified high quality dataset. I am currently building one much better than FFHQ by hand. Do you guys have a Discord or Telegram where we can collaborate?

I am looking through VGGFaceHQ today and I agree. there are many bad labels, corrupt image, general bad data. I do not like it
I have very much interest in your better dataset. I will like to collaborate.



> I agree in principle. However real time chat is easier for collaboration on details that don't have to flood this forum.

I also have interest in real time chat. May i suggest use of Matrix chat or Discord

**Comment by astalavistababe** [comment]
@GeneralShan, @somanchiu and @aesanchezgh 

1. So what if we create our own dataset with randomly using target and source generating "expected_output" (created by target + source.). 

There would be a three sets of images, target, source and "expected_output". Res could be depend for training 128, 256, 512.
Or simply use target, source and "expected_output" with 512.

2. Upscaling images with multi-upscaler like GFPGAN+GPEN2048, or other can be added like codeformer or restoreformer, etc.
![Rope-Restorer](https://github.com/user-attachments/assets/b07acfff-d808-4bc8-b9d5-4d33a43d119d)

3. Enhancing micro-skin details, for this using SD+Flux.

Ya i know this process is very computationaly and time consuming. But if we are training a model from scratch then why not train it perfectly. So it could compete with original inswapper128.

Because i saw results of rewsapper's existing model that you trained, i integrated in Rope and results are tremendously good while retaining expressions intact.

**Comment by GeneralShan** [comment]
@aesanchezgh the issue is that we are lack in ground truth. the SWAP task require GAN training, as describe in detail in SIMSWAP paper
the EXPECTED OUTPUT is ground truth that does not exist
the only way is to use EXISTING MODEL, which limit us to the existing model
possible to get minor gain from selecting BEST ONLY output from INSWAPPER, but still very limited

if we follow SIMSWAP training method, this is what I am trying
all that requires for SISMWAP training method, is dataset of many faces, and with at least MULTIPLE face of same person

this mean:
- face 1 / pic1, pic2, pic3, ... pic10

if we have dataset with many diversity of face, we can train with GAN and SIMSWAP loss function

I look at your pipeline and problem is same as VGGFace2, you run automatic restoration which introduce many artifact and failure case
if you have time possible to create dataset, please collect image of face, using public available and license data

i see big problem with VGGFace2, many image are corrupt, crop poorly, artifact, distort, wrong face, mislabel
if we have fix version it will improve model large amount

**Comment by GeneralShan** [comment]
i recommend modify your train code to use Pytorch AMP to train automatic mixed precision
it speed up significantly
when i have time later i can contribute the modification

**Comment by aesanchezgh** [comment]
> when i look through VGGFace2-HQ, it becomes more clear to me why SIMSWAP is so bad quality compare to INSWAPPER. disregarding our other identification of smaller layers

Yes exactly.

**Comment by somanchiu** [comment]
@aesanchezgh @GeneralShan
Let's chat in real-time on Discord server https://discord.com/invite/wVVPSjmj. Please don't forget to share your findings on GitHub. I spend most of my time on GitHub since my replies are slow like comments and there's no need for real-time chat. I'm not participating in the dataset discussions at the moment, as I am currently focusing on making the GAN training work.

**Comment by somanchiu** [comment]
> @GeneralShan, @somanchiu and @aesanchezgh
> 
> 2. Upscaling images with multi-upscaler like GFPGAN+GPEN2048, or other can be added like codeformer or restoreformer, etc.
>    ![Rope-Restorer](https://private-user-images.githubusercontent.com/121850606/390341446-b07acfff-d808-4bc8-b9d5-4d33a43d119d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NTQzMDcsIm5iZiI6MTczMjc1NDAwNywicGF0aCI6Ii8xMjE4NTA2MDYvMzkwMzQxNDQ2LWIwN2FjZmZmLWQ4MDgtNGJjOC1iOWQ1LTRkMzNhNDNkMTE5ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOFQwMDMzMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZTJjNDFiNzUzZDRhYTMwM2NkNDA2YzkzYjlkMTgzNzU2NmVjOWMzZjJjNjQ1ZDI3YzM0ZmNlYTAwZmZmZGNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cISZJ1QdjhxX-ew1bfzT_aBkm6YSAonLLTLYvwGh8zo)

I have tested using CodeFormer to create higher resolution images a few days ago. I trained a model based on the latest ReSwapper checkpoint for tens of thousands of steps. The model output looks like plastic, with no details, no pores, and a very smooth face. I’m not sure if this is due to the super resolution images. Intuitively, the quality should not be that low.

**Comment by GeneralShan** [comment]
@somanchiu i attempt GAN training, but i have issues. discriminator becomes perfect quickly and its loss is zero, so generator is not train (loss fluctuate)
did you observe better results?

**Comment by GeneralShan** [comment]
@astalavistababe @aesanchezgh please move dataset to external discussion

we will like to use this thread for discussion of GAN train

**Comment by GeneralShan** [comment]
I continue GAN train, I make SMALL version of the model using 512 feature size for test
I make change to train generator 20x more than discriminator., in early stage
train is more stable now

**Comment by somanchiu** [comment]
@GeneralShan I have pushed a new commit that uses the same input as InSwapper. I trained the model for 2000 steps. The model will use the target as the output and ignore the source. I have observed this behavior before when I change the input to use an image as both the target and the source in the main branch. Any ideas?

**Comment by GeneralShan** [comment]
@somanchiu is ID LOSS set correctly on generator? is discriminator HINGE LOSS set correctly?

also i think it is possible you pass incorrect batch data for reconstruct loss
simswap train loop alternate batch between SAME ID and DIFFERENT ID. SAME ID has reconstruct loss between target and output
make sure term are passing correctly and match batch type

I have not yet look at your code branch, I will try it this week

**Comment by somanchiu** [comment]
> also i think it is possible you pass incorrect batch data for reconstruct loss simswap train loop alternate batch between SAME ID and DIFFERENT ID. SAME ID has reconstruct loss between target and output make sure term are passing correctly and match batch type

@GeneralShan Thank you for the tips. I made a mistake with the reconstruction loss. I will start the model training again.

**Comment by GeneralShan** [comment]
thank you, i am glad my response is helpful

i do not have time to run my own TRAIN run this week, but i am very interested to know progress and quality of your TRAIN run

please tell me update on whether train proceed well. i have several idea to improve loss function and data if train is stable

**Comment by aesanchezgh** [comment]
> @aesanchezgh @GeneralShan Let's chat in real-time on Discord server https://discord.com/invite/wVVPSjmj. Please don't forget to share your findings on GitHub. I spend most of my time on GitHub since my replies are slow like comments and there's no need for real-time chat. I'm not participating in the dataset discussions at the moment, as I am currently focusing on making the GAN training work.

Link expired, can you resend?

**Comment by somanchiu** [comment]
A new link that never expires: https://discord.gg/vAFvCcATvc

**Comment by GeneralShan** [comment]
> The progress is not smooth. The model still uses the target as the output and ignores the source. I have removed the original data loader implemented by SimSwap because it causes high RAM usage and has some preprocessing for the images. If my understanding of the training method is correct, I have already replaced this part correctly. I will bring the original data loader back to see if I made any mistakes.

based on your report, I am still strongly suspect incorrect specification of LOSS function. normal GAN hinge loss is simple, but to train the IDENTITY TRANSFER task, the identity loss, reconstruction loss, and feature matching are the MOST essential part

Identity loss is easy, I do not doubt your implementation
Reconstruction loss is straightforward, but depend on dataloader specifying correct batch, require batch of matching identity pair of source/target and batch of non-matching identity pair. this one is EXTREMELY important
FEATURE MATCHING loss is simple specification and does not change, you can copy from original SIMSWAP code.

I will look at code later when I am having time availble

**Comment by irux** [comment]
Hey guys! I saw that Insightface released [Picsi.Ai](https://www.picsi.ai/). Do you think they still use the same architecture ?

**Comment by somanchiu** [comment]
The GAN approach now works, but it's unstable. The discriminator losses fluctuate heavily. I have committed the code to the GAN branch. Please feel free to modify the code and hyperparameters to achieve better results.
Thanks to everyone who participated in the discussion and shared your findings somewhere. Your contributions have greatly promoted the development.

![Image](https://github.com/user-attachments/assets/4ac81297-9240-44b0-9df6-e9d060fe14c4)

![Image](https://github.com/user-attachments/assets/dcdf6b34-bb4e-4b8c-bfe7-06dab863781e)

![Image](https://github.com/user-attachments/assets/2252875f-7083-4afc-899c-dd4745d01a67)

![Image](https://github.com/user-attachments/assets/7b554288-aa22-4424-8b21-b46d6196e9de)

![Image](https://github.com/user-attachments/assets/176aae51-cf2e-4d8b-b9f2-c90355b78246)

![Image](https://github.com/user-attachments/assets/1919e36a-766d-4749-b75b-032c50ccea1e)
